kable-core/src/androidMain/kotlin/Connection.kt

```diff
@@ -2,125 +2,271 @@ package com.juul.kable
+import kotlinx.coroutines.flow.filterIsInstance
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.job
+import kotlinx.coroutines.launch
 import kotlinx.coroutines.sync.Mutex
 import kotlinx.coroutines.sync.withLock
 import kotlinx.coroutines.withContext
+import kotlinx.coroutines.withTimeout
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.cancellation.CancellationException
+import kotlin.coroutines.coroutineContext
+import kotlin.reflect.KClass
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.ZERO
 
-public class OutOfOrderGattCallbackException internal constructor(
-    message: String,
-) : IllegalStateException(message)
+// Number of service discovery attempts to make if no services are discovered.
+// https://github.com/JuulLabs/kable/issues/295
+private const val DISCOVER_SERVICES_RETRIES = 5
 
 private val GattSuccess = GattStatus(GATT_SUCCESS)
 
+/**
+ * Represents a Bluetooth Low Energy connection. [Connection] should be initialized with the
+ * provided [BluetoothGatt] in a connecting or connected state. When a disconnect occurs (either by
+ * invoking [disconnect], or peripheral initiated disconnect), this [Connection] will be
+ * [disposed][close] (and cannot be re-used).
+ *
+ * To disconnect: simply call [disconnect] ([Connection] will be implicitly [closed][close] at the
+ * end of the [disconnect] sequence).
+ *
+ * If [scope], or parent [CoroutineContext] is cancelled prior to [disconnecting][disconnect], then
+ * [Connection] will be abruptly [closed][close] (upon completion of [job]) without a prior
+ * [disconnect] sequence.
+ */
 internal class Connection(
-    private val scope: CoroutineScope,
-    internal val bluetoothGatt: BluetoothGatt,
-    internal val threading: Threading,
+    parentContext: CoroutineContext,
+    internal val gatt: BluetoothGatt,
+    private val threading: Threading,
     private val callback: Callback,
+    private val services: MutableStateFlow<List<DiscoveredService>?>,
+    private val disconnectTimeout: Duration,
     logging: Logging,
 ) {
 
-    private val logger = Logger(logging, tag = "Kable/Connection", identifier = bluetoothGatt.device.address)
+    private val name = "Kable/Connection/${gatt.device}"
 
-    private val lock = Mutex()
-    private var deferredResponse: Deferred<Response>? = null
+    private val connectionJob = (parentContext.job as CompletableJob).apply {
+        invokeOnCompletion(::close)
+    }
+    private val connectionScope = CoroutineScope(
+        parentContext + connectionJob + CoroutineName(name),
+    )
+
+    val taskScope = connectionScope.childSupervisor("$name/Tasks")
+
+    private val logger =
+        Logger(logging, tag = "Kable/Connection", identifier = gatt.device.toString())
+
+    init {
+        // todo: Move this `require` to the PeripheralBuilder.
+        require(disconnectTimeout > ZERO) { "Disconnect timeout must be >0, was $disconnectTimeout" }
```

Comment:
Is this TODO dependent on some other work or can this be moved to `PeripheralBuilder` as part of this PR?

Reply:
I forget why I didn't do it in this PR. I'll try again. If I can't move it quite yet, then I'll merge w/ it as-is (with the `todo`) and try to get to it soon.

Reply:
Looked into it, I could move this, but it'd be duplicated for each platform (which is expected), but I realized that I plan to do some refactoring of `PeripheralBuilder`, so that `Config` objects are created, to make passing of those values around easier. I figured I'd just wait for that small refactor to pull this in. 🤷